### PR TITLE
Rails support

### PR DESCRIPTION
**Pull request by Guibi1**
Adds rail, powered rail, activator rail and detector rail.

## Description

Adds full rails support.
1. Allows the rails to rotate, turn and incline when placed
2. ~Implements the redstone part of rails~

Still needs to test piston interactions.

## Testing

Manual testing as of now. (i know some edge cases, would need to enforce in tests)

Please follow our [Coding Guidelines](https://github.com/Pumpkin-MC/Pumpkin/blob/master/CONTRIBUTING.md#coding-guidelines)

### CODE REVIEW

File: pumpkin/src/block/blocks/redstone/rails/rail.rs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        block: &Block,
+        _face: &BlockDirection,
+        block_pos: &BlockPos,
+        _use_item_on: &SUseItemOn,
+        player: &Player,
+        _other: bool,
+    ) -> BlockStateId {
+        let mut rail_props = RailProperties::default(block);
+
+        let shape = if let Some(east_rail) =
```
Comment (kralverde) [inline pumpkin/src/block/blocks/redstone/rails/rail.rs:40]: Is there anyway this unwrapped loop can be abstracted?

**Review by kralverde — APPROVED** [review]

File: pumpkin/src/block/blocks/redstone/rails/rail.rs (hunk 1 above)
Comment (Guibi1) [reply pumpkin/src/block/blocks/redstone/rails/rail.rs:40]: Couldn't really think of something. The behaviour is so specific it made more sense to process it that way.
Unlike the java logic, this one lazy loads the rails as needed.

**Comment by kralverde** [comment]
Can you elaborate on
>Manual testing as of now. (i know some edge cases, would need to enforce in tests)

**Comment by Guibi1** [comment]
Placement behaviour is erratic and follows the logic of what that dude in 1999 wrote (notch? lol)
Im basically asking if there's a way to unit test with blocks? If not, im not sure I can implement valid tests for that PR.
